src/IO/ReadHelpers.h

```diff
@@ -456,13 +456,14 @@ ReturnType readIntTextImpl(T & x, ReadBuffer & buf)
     }
 
 end:
-    if (has_sign && !has_number)
+    if (!has_number)
     {
-        if constexpr (throw_exception)
-            throw Exception(ErrorCodes::CANNOT_PARSE_NUMBER,
-                "Cannot parse number with a sign character but without any numeric character");
-        else
+        if constexpr (!throw_exception)
             return ReturnType(false);
+        if (has_sign)
+            throw Exception(ErrorCodes::CANNOT_PARSE_NUMBER, "Cannot parse number with a sign character but without any numeric character");
+        else
+            throw Exception(ErrorCodes::CANNOT_PARSE_NUMBER, "Cannot parse number without any numeric and sign character");
```

Comment:
`readIntTextImpl` used to return default value in `!has_sign && !has_number` case, for example when reading `try1` it returned 0 as a result. This made almost any `if (tryReadIntText(mutation, in))` check useless that's why I think it should be changed